envi/archs/aarch64/emu.py

```diff
@@ -0,0 +1,674 @@
+from envi.archs.aarch64 import A64Module
+
+logger = logging.getLogger(__name__)
+
+# CPU state (memory, regs inc SPSRs and banked registers)
+# CPU mode  (User, FIQ, IRQ, supervisor, Abort, Undefined, System)
+# 
+# instruction code
+# exception handler code
+# FIXME: SPSR handling is not certain.  
+
+# calling conventions
+
+## NO FREAKING CLUE YET....
+class A64Call(envi.CallingConvention):
+    arg_def = [(CC_REG, REG_X0), (CC_REG, REG_X1), (CC_REG, REG_X2), (CC_REG, REG_X3), 
+            (CC_REG, REG_X4), (CC_REG, REG_X5), (CC_REG, REG_X6), (CC_REG, REG_X7), 
+            (CC_STACK_INF, 8),]
+    retaddr_def = (CC_REG, REG_X30)
+    retval_def = (CC_REG, REG_X0)
+    flags = CC_CALLEE_CLEANUP
+    align = 8
+    pad = 0
+
+a64call = A64Call()
+###
+
+
+class CoProcEmulator:       # useful for prototyping, but should be subclassed
+    def __init__(self):
+        pass
+
+    def stc(self, parms):
+        logger.info("CoProcEmu: stc(%r)", parms)
+    def ldc(self, parms):
+        logger.info("CoProcEmu: ldc(%r)", parms)
+    def cdp(self, parms):
+        logger.info("CoProcEmu: cdp(%r)", parms)
+    def mcr(self, parms):
+        logger.info("CoProcEmu: mcr(%r)", parms)
+    def mcrr(self, parms):
+        logger.info("CoProcEmu: mcrr(%r)", parms)
+    def mrc(self, parms):
+        logger.info("CoProcEmu: mrc(%r)", parms)
+    def mrrc(self, parms):
+        logger.info("CoProcEmu: mrrc(%r)", parms)
+
+
+def _getRegIdx(idx, mode):
+    '''
+    Currently cheating and simply supporting only one mode and associated registers.
+    '''
+    if idx >= MAX_REGS:
+        return idx
+
+    ridx = idx
+    #ridx = idx + (mode*17)  # account for different banks of registers
+    #ridx = reg_table[ridx]  # magic pointers allowing overlapping banks of registers
+    return ridx
+
+def c0000(flags):
+    return flags & 4
+
+def c0001(flags):
+    return flags & 4 == 0
+
+def c0010(flags):
+    return flags & 2
+
+def c0011(flags):
+    return flags & 2 == 0
+
+def c0100(flags):
+    return flags & 8
+
+def c0101(flags):
+    return flags & 8 == 0
+
+def c0110(flags):
+    return flags & 1
+
+def c0111(flags):
+    return flags & 1 == 0
+
+def c1000(flags):
+    return (flags & 6) == 2
+
+def c1001(flags):
+    return (flags & c) in (0, 4, 6) # C clear or Z set
+
+def c1010(flags):
+    return (flags & 9) in (0, 9)    # N == V
+
+def c1011(flags):
+    return (flags & 9) in (1, 8)    # N != V
+
+def c1100(flags):
+    return (flags & 4) == 0 and (flags & 9) in (0, 9)   # Z==0, N==V
+
+def c1101(flags):
+    return (flags & 4) or (flags & 9) in (1, 8)         # Z==1 or N!=V (basically, "not c1100")
+
+
+conditionals = [
+        c0000,
+        c0001,
+        c0010,
+        c0011,
+        c0100,
+        c0101,
+        c0110,
+        c0111,
+        c1000,
+        c1001,
+        c1010,
+        c1011,
+        c1100,
+        c1101,
+        ]
+
+class A64Emulator(A64Module, A64RegisterContext, envi.Emulator):
```

Comment:
So there's a good number of FIXME's in this file. Are any of them things we should tackle before merge?

Reply:
done (many of them)